packages/picasso.js/src/core/data/extractor.js

```diff
@@ -18,85 +18,89 @@ export default function extract(dataConfig, data = {}, opts = {}) {
+        const extractionsConfigs = Array.isArray(dataConfig.extract) ? dataConfig.extract : [dataConfig.extract];
+        extracted.items = [];
+        const sourceFields = [];
+        extractionsConfigs.forEach((cfg) => {
+          const s = cfg.source ? data.dataset(cfg.source) : source;
+          if (!s) {
+            return;
+          }
+          extracted.items.push(...s.extract(cfg));
+          if (typeof cfg.field !== 'undefined') {
+            sourceFields.push(s.field(cfg.field));
+          }
+        });
+        if (sourceFields.length) {
+          extracted.fields = sourceFields;
         }
-        let f;
-        if (typeof obj === 'object' && typeof obj.field !== 'undefined') {
-          f = s.field(obj.field);
-        } else {
-          f = s.field(obj);
+        if (dataConfig.amend && Array.isArray(dataConfig.amend)) {
+          extracted.items.push(...dataConfig.amend);
         }
+      } else if (typeof dataConfig.field !== 'undefined' && source) {
+        const f = source.field(dataConfig.field);
         if (f) {
           if (!extracted.fields) {
             extracted.fields = [];
           }
           extracted.fields.push(f);
+          if (!('value' in dataConfig)) {
+            valueFn = f.value || (v => v);
+            labelFn = f.label || (v => v);
+            extracted.value = valueFn;
+          }
+          extracted.items = f.items().map(v => ({ value: valueFn(v), label: String(labelFn(v)), source: { field: dataConfig.field } }));
+          // TODO - add source: { key: dataConfig.source, field: dataConfig.field, data: v }
         }
-      });
-    }
+      } else if (dataConfig.fields) {
+        dataConfig.fields.forEach((obj) => {
+          const s = typeof obj === 'object' && obj.source ? data.dataset(obj.source) : source;
+          if (!s) {
+            return;
+          }
+          let f;
+          if (typeof obj === 'object' && typeof obj.field !== 'undefined') {
+            f = s.field(obj.field);
+          } else {
+            f = s.field(obj);
+          }
+          if (f) {
+            if (!extracted.fields) {
+              extracted.fields = [];
+            }
+            extracted.fields.push(f);
+          }
+        });
+      }
 
-    if (extracted.items && dataConfig.map) {
-      extracted.items = extracted.items.map(dataConfig.map);
+      if (extracted.items && dataConfig.map) {
+        extracted.items = extracted.items.map(dataConfig.map);
+      }
     }
-  }
 
-  if (dataConfig && dataConfig.stack) {
-    stack(extracted, dataConfig.stack, data.dataset);
+    if (dataConfig && dataConfig.stack) {
+      stack(extracted, dataConfig.stack, data.dataset);
+    }
   }
   if (dataConfig && !Array.isArray(dataConfig) && typeof dataConfig.sort === 'function' && extracted.items) {
     extracted.items = extracted.items.sort(dataConfig.sort);
   }
+  if (dataConfig && !Array.isArray(dataConfig) && typeof dataConfig.filter === 'function' && extracted.items) {
```

Comment:
It would probably be better performance-wise to do the filtering before sorting :)